crates/autopilot/src/run_loop.rs

```diff
@@ -529,175 +527,8 @@ impl RunLoop {
-                .checked_sub(left_sell_right_buy)
-                .unwrap_or_default();
-
-            // The difference divided by the original sell amount is the improvement in buy
-            // token. Casting to U256 is safe because the difference is smaller than the
-            // original product, which if re-divided by right.sell must fit in U256.
-            improvement
-                .checked_div(right.executed_sell.0.into())
-                .map(|v| U256::try_from(v).expect("improvement in buy fits in U256"))
-                .unwrap_or_default()
-        };
-
-        // Record best execution per order
-        let mut best_executions = HashMap::new();
-        for other in others {
-            for (uid, execution) in other.solution().orders() {
-                best_executions
-                    .entry(uid)
-                    .and_modify(|best_execution| {
-                        if !improvement_in_buy(execution, best_execution).is_zero() {
-                            *best_execution = *execution;
-                        }
-                    })
-                    .or_insert(*execution);
-            }
-        }
-
-        // Check if the solution contains an order whose execution in the
-        // solution is more than `fairness_threshold` worse than the
-        // order's best execution across all solutions
-        let unfair = solution
-            .solution()
-            .orders()
-            .iter()
-            .any(|(uid, current_execution)| {
-                let best_execution = best_executions.get(uid).expect("by construction above");
-                let improvement = improvement_in_buy(best_execution, current_execution);
-                if improvement.is_zero() {
-                    return false;
-                };
-                tracing::debug!(
-                    ?uid,
-                    ?improvement,
-                    ?best_execution,
-                    ?current_execution,
-                    "fairness check"
-                );
-                // Improvement is denominated in buy token, use buy price to normalize the
-                // difference into eth
-                let Some(order) = auction.orders.iter().find(|order| order.uid == *uid) else {
-                    // This can happen for jit orders
-                    tracing::debug!(?uid, "cannot ensure fairness, order not found in auction");
-                    return false;
-                };
-                let Some(buy_price) = auction.prices.get(&order.buy.token) else {
-                    tracing::warn!(
-                        ?order,
-                        "cannot ensure fairness, buy price not found in auction"
-                    );
-                    return false;
-                };
-                buy_price.in_eth(improvement.into()) > fairness_threshold
-            });
-        !unfair
+        let filtered_solutions = self.auction_mechanism.filter_solutions(auction, &solutions);
```

Comment:
For the combinatorial auctions it would be helpful to upload ALL submissions to the solve competitions DB collection. The submissions that got filtered out only need to be flagged as filtered or unfair.
That way it's easier to reconstruct completely how the selection procedure was executed afterwards.

Reply:
Noted. I assume this is better tackled in a follow up PR alongside the required DB changes, correct?